### PR TITLE
MàJ versions Pytorch et Pytorch-Geometric 

### PR DESCRIPTION
**Pull request by CharlesGaydon**
Checks
- [X] Pytest
- [x] Vérifier rétrocompatibilité avec précédent modèles
- [x] Versionner
- [x] Vérifier résultats d'inférence
- [x] FIx de la version de urllib3 en CICD pour Comet (qui ne passait pas dans https://github.com/IGNF/myria3d/pull/95)
- [x] Eval du gain en temps en inférence (dalle 888000_6614000) : avant MàJ 80sec d'inférence (total: 135sec), après MàJ : 82sec d'inférence (total: 131sec) -> pas de différence notable dans ce cas précis.
- [X] Eval du gain en temps en apprentissage
  - [x] Correction reset de métrique manuel maintenant nécessaire ([vérification de l'effet ici (xp privé)](https://www.comet.com/charlesgaydon/multiclass-segmentation/ba9060b0da38456d82b5882671ad5d28?experiment-tab=stdout&viewId=wiF4fXOPmCNny8a3tiAxY5Vb6))
  - [X] Correction log du chemin de l'experiment comet
- [ ] Vérification des implications de --> pour une prochaine PR
> You are using a CUDA device ('NVIDIA A40') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision

Versions de l'env installé:
image

### CODE REVIEW

**Comment by CharlesGaydon** [comment]
@MichelDaab @leavauchier Pour info, j'ai fait cette petite PR de mise à jour. Les versions n'avaient pas été updatées depuis deux ans au moins. Les tests passent en local, mais on a toujours des soucis avec la CICD - je soupconne à nouveau un problème de taille de la VM. Le problème est arrivé en changeant la  version pour urllib3, mais c'est peut-être une coïncidence : changer les requirements oblige à reconstruire une nouvelle image, et boom la mémoire sature.

Ca me semblerait bien qu'avant d'aller plus loin on trouve le problème... Histoire que les routines de déploiement soient opérationnelles quand on en a besoin.

**Comment by MichelDaab** [comment]
Es-tu sûr que le problème vient des requirements /nouvelle image / taille mémoire ?  Pour faire la modif de Léa sur l'EPSG en décembre il me semble que j'ai dû faire refaire une image aussi (j'ai souvenir que le CICD a tourné pendant un long moment une fois), et il n'y a pas eu de problème inhabituel

**Comment by CharlesGaydon** [comment]
> Es-tu sûr que le problème vient des requirements /nouvelle image / taille mémoire ? Pour faire la modif de Léa sur l'EPSG en décembre il me semble que j'ai dû faire refaire une image aussi (j'ai souvenir que le CICD a tourné pendant un long moment une fois), et il n'y a pas eu de problème inhabituel

Alors ce que je sais c'est que c'est un problème de mémoire insuffisante au moment de la création de l'environnement dans l'image docker. 

> Collecting package metadata (repodata.json): ...working... done
> Solving environment: ...working... Killed
> The command '/bin/sh -c curl -sLo /miniconda.sh https://repo.continuum.io/miniconda/Miniconda3-py39_4.10.3-Linux-x86_64.sh         && chmod +x /miniconda.sh         && /miniconda.sh -b -p /miniconda         && rm /miniconda.sh         && /miniconda/bin/conda env update -n base -f /app/environment.yml         && rm /app/environment.yml         && /miniconda/bin/conda clean -ya' **returned a non-zero code: 137**

Ce code c'est un code de mémoire excedée.
Je lis ici qu'[il est possible d'ajouter de la mémoire SWAP](https://samwize.com/2016/05/19/docker-error-returned-a-non-zero-code-137/). [Par défaut docker peut utiliser toutes les ressources du système](https://docs.docker.com/config/containers/resource_constraints/), donc utilise déjà tout le swap possible.

**Comment by CharlesGaydon** [comment]
Il n'y a pas de swap actuellement
> sudo swapon --show 
NAME      TYPE      SIZE USED PRIO
/dev/dm-1 partition 3,7G 487M   -2
Et la mémoire dure de la VM est quasi saturée. On a seulement 8Go ! Donc pas une énorme marge pour le SWAP...
> df -h /
Sys. de fichiers         Taille Utilisé Dispo Uti% Monté sur
/dev/mapper/Vol1-systeme   9,1G    7,4G  1,3G  86% /

**Comment by MichelDaab** [comment]
Peut-être que docker est sur /data, un autre disque que /dev ?

**Comment by CharlesGaydon** [comment]
> Peut-être que docker est sur /data, un autre disque que /dev ?

Ah oui probablement !
> /dev/mapper/Vol1-data                                           295G     33G  247G  12% /var/data

**Comment by CharlesGaydon** [comment]
Comparaison des temps d'apprentissage, sur 100 epochs, avec 2x3 GPUs:
image
--> Réduction de 30% du temps d'apprentissage !


|XP | Pytorch 1.11 | Pytorch 2.1|
|--|---------------------|----------------|
|Convergence | epoch 37 | epoch 44|
|Val loss | 0.106 | 0.0100|
|Test IoU | 0.690 | 0.696|


Pytorch 1.11
image
Pytorch 2.1
image

**Comment by CharlesGaydon** [comment]
La différence dans les métriques d'IoY à l'apprentissage s'explique par la MàJ de torchmetrics [qui change le comportement de `compute(...)`](https://pytorch-lightning.readthedocs.io/en/1.2.10/extensions/metrics.html) 👍🏻 
> From v1.2 onward compute() will no longer automatically call reset(), and it is up to the user to reset metrics between epochs, except in the case where the metric is directly passed to LightningModule's self.log.

Je vais donc reset les métriques manuellement, et vérifier que le comportement est bien modifié.

**Comment by CharlesGaydon** [comment]
Correctif des métrique validé : en reprenant l'entraînement après correctif (violet) les métriques sont équivalentes à celles obtenues avec la branche main (bleu)
image

File: environment.yml [hunk 1]
```diff
@@ -1,17 +1,14 @@
 # Simple install with
 # mamba env create -f environment.yml
-name: myria3d
+name: myria3d_latest_pytorch
```
Comment (leavauchier) [inline environment.yml:3]: why do you change the environment name? Is it something you used in your tests and forgot to remove?

File: environment.yml [hunk 2]
```diff
@@ -1,17 +1,14 @@
...
   - anaconda
 dependencies:
   - python==3.9.*
   - pip
-  # cudatoolkit to specify the cuda driver in the conda env
-  - conda-forge::cudatoolkit=11.3.1 # single equal sign there, not a typo
-  - numba==0.55.1
   # --------- data formats --------- #
-  - numpy==1.20
+  # - numpy
```
Comment (leavauchier) [inline environment.yml:11]: you can remove completely this line if it is not used

File: environment.yml [hunk 3]
```diff
@@ -38,26 +35,26 @@ dependencies:
   - pip:
       # --------- Deep Learning --------- #
       # Extra index may need to be on first line 
-      - --extra-index-url https://download.pytorch.org/whl/cu113
-      - torch==1.11.*
+      - --extra-index-url https://download.pytorch.org/whl/cu118
```
Comment (leavauchier) [inline environment.yml:38]: Is there a reason why you don't use conda for all these packages?
I think it is better to use conda as often as possible when you are building a conda environment to prevent issues/mismatches with dependencies

File: package_metadata.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-__version__: "3.6.1"
+__version__: "3.7.0"
```
Comment (leavauchier) [inline package_metadata.yaml:1]: si les modèles changent et ne sont pas compatibles avec les versions rpécédentes, ça vaut peut-être une version 4.0.0

Comment (CharlesGaydon) [reply package_metadata.yaml:1]: J'ai vérifié que les anciens modèles pouvaient tourner avec le nouveau code, mais je n'avais pas pensé à vérifier l'inverse. Je vais le faire :)

File: environment.yml (hunk 1 above)
Comment (CharlesGaydon) [reply environment.yml:3]: Yes indeed! Will correct.

**Comment by CharlesGaydon** [comment]
J'ai fait quelques tests de compilation de modèle (on peut encore gagner en performances, avec un facteur 1.5 à 2.5 !).
Je rencontre un soucis, j'ai ouvert une [issue sur torch_geometric](https://github.com/pyg-team/pytorch_geometric/issues/8747). Pas obligé d'ajouter cette fonctionnalité dans cette PR, on pourra séparer.

File: environment.yml (hunk 3 above)
Comment (CharlesGaydon) [reply environment.yml:35]: I changed the recipe for conda, and tested a quick training run. It seems to work fine :)

File: package_metadata.yaml (hunk 4 above)
Comment (CharlesGaydon) [reply package_metadata.yaml:1]: Ca fonctionne également, on est retro et forward compatible, donc un changement de mineur me semble suffisant :) Mais on peut tout aussi bien passer en V4.0 pour marquer le coup et savoir plus facilement quelle version est optimisée.

File: environment.yml [hunk 5]
```diff
@@ -2,33 +2,33 @@
...
-  - conda-forge::pytorch-lightning==1.5.9
-  - conda-forge::torchmetrics==0.7.*
-  - comet_ml::comet_ml==3.31.*
-  - conda-forge::urllib3<2  # To solve for https://github.com/GeneralMills/pytrends/issues/591
-  - pyg::pytorch-cluster
-  - pyg::pytorch-scatter
-  - pyg::pytorch-sparse
-  - pyg::pyg==2.1.0
+  # --------- Deep Learning --------- #
+  - pytorch=2.1
```
Comment (leavauchier) [inline environment.yml:22]: why removing the channel information? I think I added it to make sure to use the right channel for each library. But maybe there is no confusion with the newer versions

File: environment.yml [hunk 6]
```diff
@@ -2,33 +2,33 @@
...
-  - pytorch::torchvision==0.12.0
-  - conda-forge::pytorch-lightning==1.5.9
-  - conda-forge::torchmetrics==0.7.*
-  - comet_ml::comet_ml==3.31.*
-  - conda-forge::urllib3<2  # To solve for https://github.com/GeneralMills/pytrends/issues/591
-  - pyg::pytorch-cluster
-  - pyg::pytorch-scatter
-  - pyg::pytorch-sparse
-  - pyg::pyg==2.1.0
+  # --------- Deep Learning --------- #
```
Comment (leavauchier) [inline environment.yml:21]: Duplicate line

File: myria3d/metrics/iou.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    """Computes the Intersection over Union of each class in the
+    confusion matrix
+
+    Return:
+        (iou, missing_class_mask) - iou for class as well as a mask
+        highlighting existing classes
+    """
+    TP_plus_FN = confmat.sum(dim=0)
+    TP_plus_FP = confmat.sum(dim=1)
+    TP = confmat.diag()
```
Comment (leavauchier) [inline myria3d/metrics/iou.py:14]: nitpick: usually uppercase is used only for constants in python

File: myria3d/metrics/iou.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+from torch import Tensor
+
+
+def iou(confmat: Tensor):
```
Comment (leavauchier) [inline myria3d/metrics/iou.py:4]: why do you redefine iou?

File: environment.yml [hunk 9]
```diff
@@ -39,6 +39,10 @@ dependencies:
   - pandas
   - matplotlib
   # --------- loggers --------- #
+  - comet_ml=3.35
```
Comment (leavauchier) [inline environment.yml:42]: comet_ml is added twice (line 28)

File: environment.yml [hunk 10]
```diff
@@ -2,33 +2,33 @@
...
-  - pyg::pytorch-sparse
-  - pyg::pyg==2.1.0
+  # --------- Deep Learning --------- #
+  - pytorch=2.1
+  - pytorch-cuda=11.8
+  - torchvision=0.16
+  - lightning=2.0
+  - torchmetrics=0.11
+  - pyg=2.4
+  - comet_ml=3.35
```
Comment (leavauchier) [inline environment.yml:28]: no need of   - pytorch-cluster, pytorch-scatter, pytorch-sparse anymore ?

File: environment.yml [hunk 11]
```diff
@@ -2,33 +2,33 @@
 # mamba env create -f environment.yml
 name: myria3d
 channels:
+  - pytorch
+  - nvidia
+  - pyg
+  - comet_ml
   - conda-forge
   - anaconda
   - pytorch
```
Comment (leavauchier) [inline environment.yml:11]: pytorch, comet_ml, pyg added twice

File: environment.yml [hunk 12]
```diff
@@ -2,33 +2,33 @@
 # mamba env create -f environment.yml
 name: myria3d
 channels:
+  - pytorch
+  - nvidia
+  - pyg
+  - comet_ml
   - conda-forge
   - anaconda
```
Comment (leavauchier) [inline environment.yml:10]: anaconda maybe not needed?

File: environment.yml [hunk 13]
```diff
@@ -52,10 +56,7 @@ dependencies:
   - python-dotenv # loading env variables from .env file
   - rich # beautiful text formatting in terminal
   - sh # for running bash commands in some tests
-  # - scikit-learn # used in some callbacks
   - seaborn # used in some callbacks
```
Comment (leavauchier) [inline environment.yml:59]: maybe to move to the visualisation part of the file

File: environment.yml [hunk 14]
```diff
@@ -2,33 +2,33 @@
 # mamba env create -f environment.yml
```
Comment (leavauchier) [inline environment.yml:2]: I listed a few issues from the environment.yml file but I may have missed some others, I think you need to check how it was merged

Comment (CharlesGaydon) [reply environment.yml:2]: Indeed, I handled the merge poorly. Rebased and corrected the mistakes. Should be better now!

File: myria3d/metrics/iou.py (hunk 8 above)
Comment (CharlesGaydon) [reply myria3d/metrics/iou.py:4]: From MulticlassJaccardIndex we use this function `iou` to compute class-by-class IoUs. The previous method consisted of having k JaccardIndex metric object, one by class, with an inelegant modification that enabled to log single-class IoUs. 
That's actually a refactor that is not totally related to the torch upgrade :d

File: environment.yml [hunk 15]
```diff
@@ -2,35 +2,32 @@
...
+  - conda-forge
 dependencies:
-  - python==3.9.*
+  - python=3.9.*
   - pip
-  # --------- data formats --------- #
-  - numpy
-  - h5py
   # --------- Deep Learning --------- #
-  # cudatoolkit to specify the cuda driver in the conda env
```
Comment (leavauchier) [inline environment.yml:17]: no need of cuda-toolkit anymore?

File: myria3d/metrics/iou.py (hunk 8 above)
Comment (leavauchier) [reply myria3d/metrics/iou.py:6]: ok, then th'as the reason why a lot of logging stuff disappeared?
then maybe worth adding to the changelog

File: tests/myria3d/test_train_and_predict.py [hunk 16]
```diff
@@ -217,11 +216,13 @@ def _run_test_right_after_training(
     tmp_paths_overrides = _make_list_of_necesary_hydra_overrides_with_tmp_paths(
         toy_dataset_hdf5_path, tmpdir
     )
+    devices = "[0]" if accelerator == "gpu" else 1
```
Comment (leavauchier) [inline tests/myria3d/test_train_and_predict.py:219]: why enforcing it now? it looks like the previous versions could let you choose your gpu(s)

File: environment.yml (hunk 15 above)
Comment (CharlesGaydon) [reply environment.yml:17]: It is replaced with `pytorch-cuda` in later pytorch version

File: myria3d/metrics/iou.py (hunk 8 above)
Comment (CharlesGaydon) [reply myria3d/metrics/iou.py:6]: Will do !

File: tests/myria3d/test_train_and_predict.py (hunk 16 above)
Comment (CharlesGaydon) [reply tests/myria3d/test_train_and_predict.py]: In previous versions GPU 0 was always used. This is just a small refactor due to a change of signature of trainer, which requires two params: `accelerator` and `devices`, and deprecates gpus.
I made the choice to always use the same gpu to be able to run tests locally with the possibility to willingly avoid conflict with other processes (on other gpus, not on gpu 0). But I just discovered that lightning can do that automatically: 
https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#find-usable-cuda-devices
So i'll update the test.

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,9 +1,11 @@
 # CHANGELOG
 
+## 3.7.0
+- Update all versions of Pytorch, Pytorch Lightning, and Pytorch Geometric.
```
[the remark this answers is not in the document]
Comment (CharlesGaydon) [reply CHANGELOG.md:4]: done

**Comment by CharlesGaydon** [comment]
I made the requested changes @leavauchier

**Review by leavauchier — APPROVED** [review]
LGTM
